pagination/pagination.go

```diff
@@ -0,0 +1,133 @@
+
+type page struct {
+	Items      interface{} `json:"items"`
+	Count      int         `json:"count"`
+	Offset     int         `json:"offset"`
+	Limit      int         `json:"limit"`
+	TotalCount int         `json:"total_count"`
+}
+
+type Paginator struct {
+	DefaultLimit    int
+	DefaultOffset   int
+	DefaultMaxLimit int
+}
+
+func NewPaginator(defaultLimit, defaultOffset, defaultMaxLimit int) *Paginator {
+
+	return &Paginator{
+		DefaultLimit:    defaultLimit,
+		DefaultOffset:   defaultOffset,
+		DefaultMaxLimit: defaultMaxLimit,
+	}
+}
+
+func (p *Paginator) getPaginationParameters(w http.ResponseWriter, r *http.Request) (offset int, limit int, err error) {
+
+	logData := log.Data{}
+	offsetParameter := r.URL.Query().Get("offset")
+	limitParameter := r.URL.Query().Get("limit")
+
+	offset = p.DefaultOffset
+	limit = p.DefaultLimit
+
+	if offsetParameter != "" {
+		logData["offset"] = offsetParameter
+		offset, err = strconv.Atoi(offsetParameter)
+		if err != nil || offset < 0 {
+			err = errors.New("invalid query parameter")
+			log.Event(r.Context(), "invalid query parameter: offset", log.ERROR, log.Error(err), logData)
+			return 0, 0, err
+		}
+	}
+
+	if limitParameter != "" {
+		logData["limit"] = limitParameter
+		limit, err = strconv.Atoi(limitParameter)
+		if err != nil || limit < 0 {
+			err = errors.New("invalid query parameter")
+			log.Event(r.Context(), "invalid query parameter: limit", log.ERROR, log.Error(err), logData)
+			return 0, 0, err
+		}
+	}
+
+	if limit > p.DefaultMaxLimit {
+		logData["max_limit"] = p.DefaultMaxLimit
+		err = errors.New("invalid query parameter")
+		log.Event(r.Context(), "limit is greater than the maximum allowed", log.ERROR, logData)
+		return 0, 0, err
+	}
+	return
+}
+
+func renderPage(list interface{}, offset int, limit int, totalCount int) page {
+
+	return page{
+		Items:      list,
+		Count:      listLength(list),
+		Offset:     offset,
+		Limit:      limit,
+		TotalCount: totalCount,
+	}
+}
+
+func listLength(list interface{}) int {
+	l := reflect.ValueOf(list)
+	return l.Len()
+}
+
+// Paginate wraps a http endpoint to return a paginated list from the list returned by the provided function
+func (p *Paginator) Paginate(listFetcher ListFetcher) func(w http.ResponseWriter, r *http.Request) {
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		offset, limit, err := p.getPaginationParameters(w, r)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+		list, totalCount, err := listFetcher(w, r, limit, offset)
+		if err != nil {
+			return
+		}
+
+		page := renderPage(list, offset, limit, totalCount)
+
+		returnPaginatedResults(w, r, page)
+	}
+}
+
+func returnPaginatedResults(w http.ResponseWriter, r *http.Request, list page) {
+
+	logData := log.Data{}
+
+	b, err := json.Marshal(list)
+
+	if err != nil {
+		log.Event(r.Context(), "api endpoint failed to marshal resource into bytes", log.ERROR, log.Error(err), logData)
```

Comment:
I can see that the log messages have been made more generic within the paginator. The original log messages generally had the function that was being called and extra parameters in the log.data. Consider if this will negatively impact the logs without the extra context. We can use the trace ID to look at a group of messages for a request, but individually the logs will have less context.

Reply:
It is a good point, we agree that the log messages are more generic now. We have added the path to the logData to provide some more context.